src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,206 @@
+            first = newNode;
+        } else {
+            pred.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        addAllWithIndex(size, list);
+    }
+
+    public void addAllWithIndex(int index, List<T> list) {
+        Objects.checkIndex(index, size + 1);
+        Object[] a = list.toArray();
+        final int numNew = a.length;
+        Node<T> pred;
+        Node<T> succ;
+        if (index == size) {
+            succ = null;
+            pred = last;
+        } else {
+            succ = node(index);
+            pred = succ.prev;
+        }
+
+        for (Object o : a) {
+            T e = (T) o;
+            Node<T> newNode = new Node<>(pred, e, null);
+            if (pred == null) {
+                first = newNode;
+            } else {
+                pred.next = newNode;
+            }
+            pred = newNode;
+        }
+
+        if (succ == null) {
+            last = pred;
+        } else {
+            pred.next = succ;
+            succ.prev = pred;
+        }
+
+        size += numNew;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Objects.checkIndex(index,size);
+        return node(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Objects.checkIndex(index,size);
+        Node<T> x = node(index);
+        T oldVal = x.item;
+        x.item = value;
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Objects.checkIndex(index,size);
+        return unlink(node(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> x = first; x != null; x = x.next) {
+                if (x.item == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> x = first; x != null; x = x.next) {
+                if (object.equals(x.item)) {
+                    unlink(x);
+                    return true;
+
+                }
+            }
+        }
         return false;
     }
 
+    private T unlink(Node<T> x) {
+        final T element = x.item;
+        Node<T> next = x.next;
+        Node<T> prev = x.prev;
+
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            x.prev = null;
+        }
+
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            x.next = null;
+        }
+
+        x.item = null;
+        size--;
+        return element;
+    }
+
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        if (first == null) {
+            return true;
+        } else {
+            return false;
+        }
```

Comment:
```suggestion
	return size == 0; 
```

Мы можем сделать это проще